From 0aad5348596d709e402cdbbd6252a6c0ba05b71e Mon Sep 17 00:00:00 2001
From: Colin McDonald <macguy8.main@gmail.com>
Date: Fri, 10 Jul 2015 17:24:13 -0400
Subject: [PATCH] Optimize getCubes


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 14f13fc17..16eb99ccb 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -1312,7 +1312,7 @@ public abstract class World implements IBlockAccess {
         }
         // Spigot end
 
-        double d0 = 0.25D;
+        /*double d0 = 0.25D;
         List list = this.getEntities(entity, axisalignedbb.grow(d0, d0, d0));
 
         for (int j2 = 0; j2 < list.size(); ++j2) {
@@ -1326,7 +1326,7 @@ public abstract class World implements IBlockAccess {
             if (axisalignedbb1 != null && axisalignedbb1.b(axisalignedbb)) {
                 this.L.add(axisalignedbb1);
             }
-        }
+        }*/
 
         return this.L;
     }
-- 
2.13.3

